### PR TITLE
generate spec navigation

### LINKED ISSUE #176: mkdocs watcher does not pick up navigation changes
Currently, when you are editing markdown files under `./documenation` you can run `documentation/scripts/watch.sh` in parallel to just run a local version of the website, and you can see your changes live.

But due to how we generate part of the specification now, we also generate a `generated/mkdocs.navigation.yml` file, and combine it with the hand-authored root `mkdocs.config.yml` file in a pre-build step. This means that `watch.sh` doesn't pick up changes to the original navigation YAML like it does with markdown.

After #153, it is now trivial/straightforward to generate and verify the structure from the `spec` codegen crate. I will do that and remove the extra build step, so that the watcher can pick up the changes to `mkdocs.config.yml` navigation section as well. I will also make it resilient to any missing files/build breaks, and restart again when they are fixed.

The current workaround is to just rerun `documentation/scripts/watch.sh` (near-instant reload) whenever you hand-edit the root `mkdocs.config.yml`. Any changes to markdown should still be picked up correctly.

### PR DESCRIPTION
**Pull request by OmarTawfik**
- Build changes required to start generating a complete spec of each language version.
- Moves spec contents to respective crates/solidity/inputs and crates/solidity/outputs directories.
- Greatly simplifies the spec build process/code, by specifying directory structures in separate files instead of composing a custom mkdocs.yml at build time.
- Generates each production in a single snippet file, that is imported at build time into respective spec pages.
- Produce smaller diffs in code review, instead of changing every file that imports that rule. Otherwise, it will be much worse in the next PR when we start generating a single page per grammar version.
- Pass name of parent productions when writing local definitions, which is needed to generate the correct links in spec. So I added that and fixed callsites.
- Generate pre-formatted snippets, instead of repeatedly calling `prettier`, which brought down our rebuild time for dev REPL from 11s to 5s.
- Fixes #176

### CODE REVIEW

File: crates/codegen/spec/src/public/mod.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    grammar: &'a Grammar,
+    snippets: &'a Snippets<'a>,
+}
+
+impl<'a> PublicPages<'a> {
+    pub fn new(grammar: &'a Grammar, snippets: &'a Snippets<'a>) -> Self {
+        return Self { grammar, snippets };
+    }
+
+    pub fn write_to_disk(
```
Comment (AntonyBlakey) [inline crates/codegen/spec/src/public/mod.rs:24]: Not so keen on this name ... `disk` ? Maybe not a disk ...

File: crates/codegen/spec/src/public/navigation.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    },
+    Page {
+        title: &'a str,
+        path: &'a str,
+        contents: String,
+    },
+}
+
+impl<'a> NavigationEntry<'a> {
+    pub fn write_to_disk(&self, codegen: &mut CodegenContext, parent_dir: &PathBuf) -> Result<()> {
```
Comment (AntonyBlakey) [inline crates/codegen/spec/src/public/navigation.rs:22]: Not so keen on this name ... `disk` ? Maybe not a disk ...

**Review by AntonyBlakey — APPROVED** [review]
2 comments

File: crates/codegen/spec/src/public/navigation.rs (hunk 2 above)
Comment (OmarTawfik) [reply crates/codegen/spec/src/public/navigation.rs]: Changed to `write_files`. How about that?

File: crates/codegen/spec/src/public/mod.rs (hunk 1 above)
Comment (OmarTawfik) [reply crates/codegen/spec/src/public/mod.rs]: Changed to `write_files`. How about that?

**Pull request merged**